### ISSUE TITLE
deadlock in SyncChunkSaveState

### ISSUE THREAD

**Issue by shanson7** [grafana/metrictank#726]
We had Cassandra slowness which caused the write queues to fill in MT, but we noticed that on many of our instances, one worker never drained. Over the course of a couple minutes this caused all ingest in MT to stop.

Attaching a snapshot of the dashboard. It's a bit hard to tell, so I added an arrow showing where queue 5 just hangs as the other queues all start to drain. Also attached a stack from one of our hung instances.

![screen shot 2017-09-11 at 5 04 43 pm](https://user-images.githubusercontent.com/5024860/30329346-0921efe2-97a0-11e7-8bf1-79308d36ee85.jpg)

[metrictank.20170911.stack.txt](https://github.com/raintank/metrictank/files/1296155/metrictank.20170911.stack.txt)

**Comment by Dieterbe** [comment]
can you post an interactive snapshot of this chart (see grafana dashboard export menu), you can send the link to me in private if you want.

**Comment by shanson7** [comment]
Nifty: https://snapshot.raintank.io/dashboard/snapshot/LDKE6Tlw1g7Zl3MGhN9r7mRwmc3PS6o9?orgId=2

**Comment by Dieterbe** [comment]
looks like you meant queue 2, not queue 5.
i'm sure i have once experienced the exact same thing in a local dev stack, but i couldn't reproduce it.
i couldn't find the right ticket, maybe #539 is a bit related.

**Comment by shanson7** [comment]
Sorry, yeah, the pic, stack and interactive snapshot are from different instances.

We can reliably reproduce (in fact, we are having trouble not reproducing it).

We did notice that `nodetool tpstat` showed that `Native-Transport-Requests` were blocked. Although, once that stat went back to 0, metrictank stayed stuck. Even bouncing all cassandra instances didn't help.

**Comment by shanson7** [comment]
After hitting this again, I think it may be related to https://github.com/gocql/gocql/issues/928

**Comment by Dieterbe** [comment]
do your MT instances log any errors (or warnings) ? anything related to cassandra/gocql/reconnects? or query errors?

I don't think this is related to that gocql issue, it talks about concurrent use of `GetHosts()` in gocql, but that function doesn't show up in the stacktrace you posted.

looking at the stacktrace, and combing through them with [my patched version of panicparse](https://github.com/maruel/panicparse/pull/20) shows a whole lot of routines stuck in netpoll (gocql.readHeader) - nothing strange here I think - and then a couple of routines stuck in select calls:

```
cat metrictank.20170911.stack.txt | panicparse -filter-package gocql  | tail -n 18
1: IO wait [3603 minutes] [Created by gocql.Connect @ conn.go:259]
    net   netpoll.go:164        runtime_pollWait(#1889, 0x72, 0x61)
    net   fd_poll_runtime.go:75 (*pollDesc).wait(#1408, 0x72, #23, #19)
    net   fd_poll_runtime.go:80 (*pollDesc).waitRead(#1408, #1817, 0x1000)
    net   fd_unix.go:250        (*netFD).Read(#1407, #1817, 0x1000, 0x1000, 0, #23, #19)
    net   net.go:181            (*conn).Read(#1657, #1817, 0x1000, 0x1000, 0, 0, 0)
    bufio bufio.go:213          (*Reader).Read(#1336, #1036, 0x1, 0x9, #1890, 0, 0x45a8f6)
    io    io.go:307             ReadAtLeast(#20, #1336, #1036, 0x1, 0x9, 0x1, 0x30, 0x28, 0xc13080)
    io    io.go:325             ReadFull(#20, #1336, #1036, 0x1, 0x9, 0, #2195, 0)
    gocql frame.go:360          readHeader(#20, #1336, #1036, 0x9, 0x9, 0, 0, 0, 0, 0, ...)
    gocql conn.go:466           (*Conn).recv(#1035, 0, 0)
    gocql conn.go:428           (*Conn).serve(#1035)
2: select [Created by gocql.(*Session).init @ session.go:202]
    gocql session.go:224        (*Session).reconnectDownedHosts(*, #75)
4: select [3428~3884 minutes] [Created by gocql.newEventDebouncer @ events.go:28]
    gocql events.go:40          (*eventDebouncer).flusher(*)
2: select [Created by gocql.(*controlConn).connect @ control.go:235]
    gocql control.go:63         (*controlConn).heartBeat(*)
```

However all 3 of those selects look benign to me. they're just waiting for some timer/ticker/timeout event.
I don't see evidence of a deadlock in gocql.

BTW, based on the stacktrace, I see that you're using a version of MT before 0.7.4-99-ga04083f, 
as that commit brought a change to the imported gocql code.
what version are you running exactly?
you may want to update to 0.7.4-99-ga04083f, or latest master even. the only thing to note is that since 
#685 the read queue behavior went from blocking to non-blocking, so you'll need to set larger read queue values to be able to accomodate temporary pile-ups of bursts of reads (basically look at your high values for gets/s and then multiply by how many seconds of pile-up you want to tolerate, i suggest 20~40 seconds, and just multiply those numbers to get a good queue size)

**Comment by shanson7** [comment]
I think I found it. 

Entertain this theory:

1. [Thread 1]  A particular chunk has it's raw data and it's aggregations that need persisting
2. [Thread 1] The raw chunk is successfully pushed on the queue.
3. [Thread 1] When adding the aggregations, the queue fills causing a blocking push (whilst holding the raw AggMetric lock)
4. [Thread 2] The raw chunk has been successfully saved to cassandra and it attempts to save the chunk state (which locks on the same lcok as Thread 1)

Thread 1 will never relinquish the lock, since it is blocked on the queue. The queue won't drain, since the worker (Thread 2) is blocked on Thread 1.

Thread 1:
```
1: chan send [1224 minutes] [Created by kafkamdm.(*KafkaMdm).Start @ kafkamdm.go:216]
    mdata         store_cassandra.go:318 (*CassandraStore).Add(#2547, #3616)
    mdata         aggmetric.go:421       (*AggMetric).persist(#5283, 0)
    mdata         aggmetric.go:510       (*AggMetric).add(#5283, #67, #9097)
    mdata         aggmetric.go:435       (*AggMetric).Add(#5283, #9072, #9097)
    mdata         aggregator.go:83       (*Aggregator).flush(#4356)
    mdata         aggregator.go:107      (*Aggregator).Add(#4356, #74, #9096)
    mdata         aggmetric.go:330       (*AggMetric).addAggregators(#5280, #3292, #9096)
    mdata         aggmetric.go:537       (*AggMetric).add(#5280, #74, #9096)
    mdata         aggmetric.go:435       (*AggMetric).Add(#5280, #5925, #9096)
```

Thread 2:
```
5: semacquire [1223~1224 minutes] [Created by mdata.NewCassandraStore @ store_cassandra.go:297]
    sync          sema.go:62             runtime_SemacquireMutex(*)
    sync          mutex.go:87            (*Mutex).Lock(*)
    sync          rwmutex.go:86          (*RWMutex).Lock(*)
    mdata         aggmetric.go:78        (*AggMetric).SyncChunkSaveState(*, *)
    mdata         store_cassandra.go:343 (*CassandraStore).processWriteQueue(#2547, *, *)
```

I'm not sure why the same

**Comment by shanson7** [comment]
It seems like the comment at https://github.com/grafana/metrictank/blob/116c2f7b707b9bfcadf514d48599357aa05eba45/mdata/aggmetric.go#L324 and the one at https://github.com/grafana/metrictank/blob/116c2f7b707b9bfcadf514d48599357aa05eba45/mdata/aggmetric.go#L141 seem to disagree with each other.

It we can simply unlock before calling `addAggregators` this deadlock should go away.

**Comment by shanson7** [comment]
As an FYI, I rolled out a change that simply does this:
```go
	// Temp hack/workaround for https://github.com/grafana/metrictank/issues/726
	a.Unlock()
	a.addAggregators(ts, val)
	a.Lock()
```

to add the datapoint (and possibly cause the persist action) outside of the parent lock. It has been running in our cluster for 2 days without any hangs (or other issues). 

I think another possible solution is to always call `addAggregators` before `persist`, but that's a bit more "social contract" than I would like.

None of these options feel very clean.

**Comment by Dieterbe** [comment]
this is a bit hard to follow. you mention store_cassandra.go:297 and store_cassandra.go:318 in your thread theory examples, but i don't see those in the stacktrace you shared. can you please confirm which version you were running when you got that stacktrace, and which exact version you were running as you posted these snippets, so that i can check lines 297 and 318.

that said, your theory does seem to make sense.  please confirm your versions used and i'll look more into it.

as far as the comments that seem to disagree, addAggregators refers to the non-thread safety of adding points to aggregators, hence needing to hold a lock. while GetAggregated refers to reading the slice of aggregators (the `aggregators` attribute of AggMetric) which doesn't change once the initial setup at creation-time is done, so is safe to read/iterate concurrently. (note that that function, as it iterates over aggregators, calls the Get() of their internal AggMetric structures, which has its own lock)

I think your patch happens to look safe because there is sufficient time between adding points to the same AggMetric, which in all environments is generally true, until it is not. once a bunch of points are being added to the same AggMetric concurrently, the adds to the AggMetric itself will be safe due to the lock, but the adds to its aggregators will be unsafe.

once i have the version i'll look into this more and try to find a clean fix.

**Comment by shanson7** [comment]
[20171020.goroutines.txt](https://github.com/grafana/metrictank/files/1415822/20171020.goroutines.txt)
Ah, I'm sorry. I was working off of another stack trace after a rebuild from master (commit hash d2c63ec84920310ab58a4abc72078dd2a2e36388).


As far as the thread safety, I mistakenly though that the `Aggregator` had a lock, but only the underlying `AggMetric`s do, so shared state like `currentBoundary` is unprotected.

**Comment by Dieterbe** [comment]
> shared state like currentBoundary is unprotected.

to avoid any confusion; it is protected by the lock on the parent AggMetric.  those fields are only modified during an Aggregator.Add call (or Aggregator.flush but that's only called by the former), and the Add method is only called from AggMetric.addAggregators which is in the critical section of the AggMetric lock.
(it is however unprotected with your patch applied)

so describing the problem with slightly better terminology:

CassandraStore.Add does not return until the cwr can be put on the queue, keeping persist blocking and thus holding the AggMetric lock.  we can only drain the queue when we can call AggMetric.SyncChunkSaveState which needs the AggMetric lock.

your explanation is a case of this going wrong (aggregators trying to flush while their parent AggMetric is locked, but on a queue full they block and hold the lock, so that SyncChunkSaveState can't work either)

but do we even need the Aggregators active here to run into this issue? consider the case where we call AggMetric.Add() (without any aggregators) or AggMetric.GC() runs .
both of these call persist. Note that persist may trigger multiple chunk saves.
(though admittedly this wouldn't really happen with the way you and us deploy MT.  this only happens if you have a secondary node with a bunch of unsaved chunks that suddenly gets promoted to primary)
we can also get into a situation here where a chunk gets saved and tries to call SyncChunkSaveState but then the queue fills up and the other persist calls hang, keeping the lock open and blocking the SyncChunkSaveState, and hence further progress.

I've never been a big fan of the SyncChunkSaveState stuff, i always thought it was a weird cyclic messaging approach. perhaps the store should at least put a `go ` in front of that call so that it doesn't block further processing, though that will create a large amount of goroutines, which come with some overhead. the store could also keep a `map[*AggMetric]ts` which the aggmetrics can then periodically query for. not pretty though.  We can also perhaps decouple the Aggregator more from its parent AggMetric, and do something akin to your patch, but add locking in each Aggregator.

> I think another possible solution is to always call addAggregators before persist

doesn't this result in the same problem? e.g. addAggregators causes aggregators flushing, some of them can put cwr's in the write queue, others are blocked from putting stuff in the queue, so the AggMetric lock would still be held, so SyncChunkSaveState can't get the lock and we get the same deadlock.

**Comment by shanson7** [comment]
> (it is however unprotected with your patch applied)

Yes, I was referring to the thread (un)safety of the hacky workaround.

> but do we even need the Aggregators active here to run into this issue? 

Any case where >1 cwr's are enqueued under the lock required to SyncChunkSaveState is subject to this issue. 

Calling addAggregators before persist would get around the original case iff each aggregator persisted <=1 cwr (the common case, I'd wager). This is because SyncChunkSaveState would only need the lock on the child AggMetric which would have been relinquished after Add completed. 

However, this is basically just reducing the chances of the issue occurring.

**Comment by shanson7** [comment]
Is it possible to build the set of `cwr`s under the lock and enqueued them outside of the lock?

**Comment by Dieterbe** [comment]
now that your queue sizes are sized better, this isn't as urgent anymore right?
finding the proper solution requires some more time

**Comment by shanson7** [comment]
Correct. This is only an issue if the queue fills up, which shouldn't happen under normal circumstances.

**Comment by woodsaj** [comment]
We are still seeing this happen occasionally.

To me this definitely looks like a gocql bug as the CassandraStore.processWriteQueue goroutine is blocked waiting for gocql to respond.
```
goroutine 796 [select, 2231 minutes]:
github.com/raintank/metrictank/vendor/github.com/gocql/gocql.(*Conn).exec(0xc420a9e000, 0x0, 0x0, 0x1025720, 0xc4345e1f10, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/ubuntu/.go_workspace/src/github.com/raintank/metrictank/vendor/github.com/gocql/gocql/conn.go:637 +0x74f
github.com/raintank/metrictank/vendor/github.com/gocql/gocql.(*Conn).executeQuery(0xc420a9e000, 0xc430732200, 0xc407cc7dfe)
	/home/ubuntu/.go_workspace/src/github.com/raintank/metrictank/vendor/github.com/gocql/gocql/conn.go:830 +0x575
github.com/raintank/metrictank/vendor/github.com/gocql/gocql.(*Query).execute(0xc430732200, 0xc420a9e000, 0x1069b20)
	/home/ubuntu/.go_workspace/src/github.com/raintank/metrictank/vendor/github.com/gocql/gocql/session.go:750 +0x35
github.com/raintank/metrictank/vendor/github.com/gocql/gocql.(*queryExecutor).executeQuery(0xc420986180, 0x1033660, 0xc430732200, 0xc42f7a0013, 0x3, 0xc427de7c68)
	/home/ubuntu/.go_workspace/src/github.com/raintank/metrictank/vendor/github.com/gocql/gocql/query_executor.go:42 +0x1b3
github.com/raintank/metrictank/vendor/github.com/gocql/gocql.(*Session).executeQuery(0xc420184280, 0xc430732200, 0xc427de7c00)
	/home/ubuntu/.go_workspace/src/github.com/raintank/metrictank/vendor/github.com/gocql/gocql/session.go:376 +0xc1
github.com/raintank/metrictank/vendor/github.com/gocql/gocql.(*Query).Iter(0xc430732200, 0xc430732200)
	/home/ubuntu/.go_workspace/src/github.com/raintank/metrictank/vendor/github.com/gocql/gocql/session.go:921 +0xbf
github.com/raintank/metrictank/vendor/github.com/gocql/gocql.(*Query).Exec(0xc430732200, 0xc42b09f1d0, 0x45)
	/home/ubuntu/.go_workspace/src/github.com/raintank/metrictank/vendor/github.com/gocql/gocql/session.go:904 +0x2b
github.com/raintank/metrictank/mdata.(*CassandraStore).insertChunk(0xc42089a600, 0xc4240dcf30, 0x26, 0xa8c005a2627d0, 0xc43c631520, 0x4a, 0xc8, 0x0, 0x0)
	/home/ubuntu/.go_workspace/src/github.com/raintank/metrictank/mdata/store_cassandra.go:370 +0x467
github.com/raintank/metrictank/mdata.(*CassandraStore).processWriteQueue(0xc42089a600, 0xc420a881e0, 0xc420bcc0e0)
	/home/ubuntu/.go_workspace/src/github.com/raintank/metrictank/mdata/store_cassandra.go:309 +0x398
created by github.com/raintank/metrictank/mdata.NewCassandraStore
	/home/ubuntu/.go_workspace/src/github.com/raintank/metrictank/mdata/store_cassandra.go:267 +0x1008
```

the select that is blocking should not be held for so long.  Even if the resp never arrives, the timeoutCh or ctxDone channels should fire.

**Comment by woodsaj** [comment]
So i think we definitely need to update gocql if we haven't already, another deadlock fix was merged a month ago.

As an additional layer of protection, we should also query with a context.WithTimeout

**Comment by woodsaj** [comment]
Looking at @shanson7 stacks, he definitely is seeing a deadlock in SyncChunkSaveState().

So looks like we have 2 issues that have the same symptoms.  I going to move the gocql deadlock problem to a separate issue.

**Comment by shanson7** [comment]
This hit us rather unexpectedly again on Wednesday. Still early debugging but we had an unexpected jump in CWR that hit what we though was a "safe" queue size. :/

**Comment by Dieterbe** [comment]
we just merged the fix for this.
